### PR TITLE
Fixes: #3030 - Truncate the logs from the correct range

### PR DESCRIPTION
**Pull request by BramGruneir**
instead of always truncating the logs of the first range.
Also add in a way to double check that the range being truncated is
the correct one by requiring that the rangeID is included in the request.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3097)

### CODE REVIEW

File: roachpb/api.proto [hunk 1]
```diff
@@ -515,6 +515,12 @@ message TruncateLogRequest {
 
   // Log entries < this index are to be discarded.
   optional uint64 index = 2 [(gogoproto.nullable) = false];
+
+  // Range ID to double check that the correct range is being truncated. The
+  // range may have changed form the one specified in the span in the case of
```
Comment (bdarnell) [inline roachpb/api.proto]: s/span/header/

**Comment by bdarnell** [comment]
LGTM

File: roachpb/api.proto (hunk 1 above)
Comment (tschottdorf) [inline roachpb/api.proto]: s/form/from/

File: roachpb/api.proto [hunk 2]
```diff
@@ -515,6 +515,12 @@ message TruncateLogRequest {
...
+  // a merge.
+  optional int64 range_id = 6 [(gogoproto.nullable) = false,
```
Comment (tschottdorf) [inline roachpb/api.proto]: s/6/3/

File: storage/replica_command.go [hunk 3]
```diff
@@ -1049,7 +1062,6 @@ func (r *Replica) TruncateLog(batch engine.Engine, ms *engine.MVCCStats, h roach
 	if err != nil {
 		return reply, err
 	}
-	rangeID := r.Desc().RangeID
 	start := keys.RaftLogKey(rangeID, 0)
 	end := keys.RaftLogKey(rangeID, args.Index)
 	if err = batch.Iterate(engine.MVCCEncodeKey(start), engine.MVCCEncodeKey(end), func(kv roachpb.RawKeyValue) (bool, error) {
```
Comment (tschottdorf) [inline storage/replica_command.go]: seems like `ClearRange` would be useful here.

**Comment by tschottdorf** [comment]
LGTM

**Comment by tamird** [comment]
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[roachpb/api.proto, line 520 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3097#-K2nJsSZW10Pb3HJ1IYI-r1-520)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2c3e2e06d04a8d97934f04c39e5f410d19e1a673/roachpb/api.proto#L520)):</sup>
The whole comment elides some detail. In particular, it forces the reader to look at the definition of Span to convince themselves that Span does not in fact carry a RangeID. Can you rewrite this to be a bit more cache friendly?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3097)

File: roachpb/api.proto (hunk 1 above)
Comment (BramGruneir) [inline roachpb/api.proto]: fixed and fixed

File: roachpb/api.proto (hunk 2 above)
Comment (BramGruneir) [inline roachpb/api.proto]: fixed

File: storage/replica_command.go (hunk 3 above)
Comment (BramGruneir) [inline storage/replica_command.go]: Might be, not going to change this up now.

**Pull request merged**